gridfs.go

```diff
@@ -0,0 +1,601 @@
+		"operation":     "put",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "resolving bucket")
+	}
+
+	if b.opts.DryRun {
+		return nil
+	}
+
+	if _, err = grid.UploadFromStream(b.normalizeKey(name), input); err != nil {
+		return errors.Wrap(err, "uploading file")
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "get",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	return b.Reader(ctx, name)
+}
+
+func (b *gridfsBucket) Upload(ctx context.Context, name, path string) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"dry_run":       b.opts.DryRun,
+		"operation":     "upload",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+		"path":          path,
+	})
+
+	f, err := os.Open(path)
+	if err != nil {
+		return errors.Wrapf(err, "opening file '%s'", name)
+	}
+	defer f.Close()
+
+	return errors.WithStack(b.Put(ctx, name, f))
+}
+
+func (b *gridfsBucket) Download(ctx context.Context, name, path string) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "download",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+		"path":          path,
+	})
+
+	reader, err := b.Reader(ctx, name)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err = os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+		return errors.Wrapf(err, "creating enclosing directory for file '%s'", path)
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return errors.Wrapf(err, "creating file '%s'", path)
+	}
+	_, err = io.Copy(f, reader)
+	if err != nil {
+		_ = f.Close()
+		return errors.Wrap(err, "copying data")
+	}
+
+	return errors.WithStack(f.Close())
+}
+
+func (b *gridfsBucket) Push(ctx context.Context, opts SyncOptions) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"dry_run":       b.opts.DryRun,
+		"operation":     "push",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"remote":        opts.Remote,
+		"local":         opts.Local,
+		"exclude":       opts.Exclude,
+	})
+
+	var re *regexp.Regexp
+	var err error
+	if opts.Exclude != "" {
+		re, err = regexp.Compile(opts.Exclude)
+		if err != nil {
+			return errors.Wrap(err, "compiling exclude regex")
+		}
+	}
+
+	localPaths, err := walkLocalTree(ctx, opts.Local)
+	if err != nil {
+		return errors.Wrap(err, "finding local paths")
+	}
+
+	for _, path := range localPaths {
+		if re != nil && re.MatchString(path) {
+			continue
+		}
+
+		target := b.Join(opts.Remote, path)
+		_ = b.Remove(ctx, target)
```

Comment:
Is this an ignorable error? If it fails, will the upload fail?

Reply:
I will add a comment but I do not think it is worth an error. For consistent behavior across our bucket implementations we want to "overwrite" files. GridFS will keep a copy of the old file but fetching by filename [defaults to the most recent revision](https://github.com/mongodb/mongo-go-driver/blob/v1/mongo/options/gridfsoptions.go#L174C1-L187). Likewise, uploading by filename will create a new revision (if you specify the unique ID of the revision you can then append but we don't want that to happen so we don't expose that behavior in via pail). 

Removing is good because it can save space but does need to impact behavior.

Reply:
Actually, this behavior can get tricky. What if the upload operation fails and we remove the previous revision? I am going to remove it and keep around old revisions. This isn't going to impact much since no one uses the gridfs impl except for testing anyways.

Reply:
Gotcha, I wasn't clear that the remove was trying to remove old file revisions. Keeping the old revisions SGTM.